# PhD-Thesis
Data and scripts from my thesis

---
## Chapter 2 Probabilistic models that use k-mer information for short-read genotyping in repetitive polymorphic genomic regions
- Figure 2.3: Haploid allele calling performance of the count-coverage model.
- Figure 2.4: Per-allele haploid calling performance of the count-coverage model.
- Figure 2.5: Effect of read and fragment length on the count-coverage model performance.
- Figure 2.6: Comparison of ùúÜ estimation methods used in the count model.
- Figure 2.7: Comparison of different methods for determining allele likelihoods.
- Figure 2.8: Diploid genotype calling performance of the count-coverage model.
- Figure 2.10: Haploid allele calling performance of the distance model. 
- Figure 2.11: Haploid allele and diploid genotype calling performance of the distance-max model compared to the count-coverage model.
- Figure 2.12: Haploid allele calling performance of combining the count and distance models. 
- Figure 2.13: Combining the count and distance models for genotyping diploid simulations.
- Figure 2.15: Assessing different methods of calculating ùúÜ for the count model on real sequencing data. 
- Figure 2.16: Distribution of k-mer coverage across the PRDM9 zinc finger and flanking regions. (A) 
- Figure 2.17: Effect of GC content across the zinc finger region.
- Figure 2.18: The ratio of k-mer counts to per-base read coverage for HG003 short and long reads. 
- Figure 2.19: The effect of correcting reads on the distribution of k-mer coverage across the zinc finger region for HG003.
- Figure 2.21: The effect of correcting sequencing reads prior to genotyping real sequencing data. 
- Figure 2.22: Diploid calling performance of the count-coverage model on simulated A/A and A/L37 genotypes using a range of manually specified ùúÜ values.
- Figure 2.23: Effect of read correction on genotyping the Ashkenazi trio with the count-coverage model using a range of manually specified ùúÜ values.


## Chapter 3 A haplotype-based model for long-read genotyping and validation of the short-read genotyping models
- Figure 3.4: Performance of the k-mer count models on HPRC++ and OHS long-read data.
- Figure 3.6: Effect of the known allele list size on the short-read genotyping model calls. 
- Figure 3.7: Clustering of OHS samples by inferred ancestry.

## Chapter 4 Improving read alignment in polymorphic genomic regions using graph-based reference genomes
- Figure 4.2: Short path accuracy for different PRDM9 reference graph topologies.
- Figure 4.3: Performance of different paired-end read treatments during alignment by vg.
- Figure 4.4: Spurious improvements and impairments in read alignments to the different PRDM9 graph topologies.
- Figure 4.5: Effect of embedded paths in read alignments to the PRDM9 reference graph topologies. 
- Figure 4.6: True alignment improvements to the different PRDM9 graph topologies. 
- Figure 4.7: True alignment improvements to the different PRDM9 graph topologies in the zinc finger region alone.
- Figure 4.8: Improvement in read alignments for the HPRC++ samples using a PRDM9 reference graph compared to GRCh38.
